runtime/protos/runtime.proto

```diff
@@ -17,232 +17,330 @@ option go_package = "github.com/ServiceWeaver/weaver/runtime/protos";
+// a *handshake*. The envelope sends an EnvelopeInfo, and the weavelet responds
+// with a WeaveletInfo.
+//
+//                   ENVELOPE                  WEAVELET
+//                   ───┬────                  ───┬────
+//                      │ ──────────────────────> │ EnvelopeInfo
+//                      │ <────────────────────── │ WeaveletInfo
+//                      :                         :
+//
+// After the handshake, the envelope and weavelet communicate freely. There are
+// three communication patterns.
+//
+//     1. The envelope can initiate an RPC against the weavelet.
+//     2. The weavelet can initiate an RPC against the envelope.
+//     3. The weavelet can send messages to the envelope (without any replies).
+//
+// Here is an example of these three types of communication.
+//
+//                   ENVELOPE                  WEAVELET
+//                   ───┬────                  ───┬────
+//                   1. │ ──────────────────────> │ UpdateComponentsRequest
+//                      │ <────────────────────── │ UpdateComponentsReply
+//                      │                         │
+//                   2. │ <────────────────────── │ GetListenerAddressRequest
+//                      │ ──────────────────────> │ GetListenerAddressReply
+//                      │                         │
+//                   3. │ <────────────────────── │ LogEntry
+//                      │ <────────────────────── │ TraceSpans
+//                      :                         :
+//
+// Note that these three types of communication can intersperse.
+//
+//                   ENVELOPE                  WEAVELET
+//                   ───┬────                  ───┬────
+//                      │ <────────────────────── │ EnsureComponentRequest
+//                      │ ──────────────────────> │ UpdateComponentsRequest
+//                      │ <────────────────────── │ UpdateComponentsReply
+//                      │ <────────────────────── │ LogEntry
+//                      │ ──────────────────────> │ UpdateRoutingInfoRequest
+//                      │ <────────────────────── │ UpdateRoutingInfoReply
+//                      │ ──────────────────────> │ EnsureComponentReply
+//                      :                         :
+//
+// To avoid deadlocks, a weavelet is guaranteed to process an RPC without
+// blocking and without touching the pipe.
+
+// EnvelopeMsg is a message sent by an envelope to a weavelet.
 message EnvelopeMsg {
-  // Unique id identifies a request that is part of a synchronous call. A
-  // request with positive id x expects a reply with negative id -x.
-  // Non-synchronous calls have an id of 0.
+  // TODO(mwhittaker): Don't make UpdateRoutingInfoRequest and
+  // UpdateComponentsRequest RPCs? They can be one-directional messages like
+  // logs and traces.
+
+  // A unique id for synchronous calls. A request with positive id x expects a
```

Comment:
Should we talk about synchronous/non-synchronous calls in the documentation? Are these RPC vs non RPC calls? It would help to clarify the terminology.

Reply:
Also, maybe an example how this could be used would help (because is something the developer of the deployer has to set?).